src/main/java/com/google/starfish/services/FavoriteNoteService.java

```diff
@@ -115,6 +128,141 @@ public long getNumFavoritesByNoteId(DataSource pool, long noteId) throws SQLExce
+    return getTrendingNotes(pool, Recency.ALL_TIME);
+  }
+
+  /** Gets trending notes today filtered by school or course */
+  public Object[][] getTrendingNotesTodayBySchoolOrCourse(DataSource pool, String school, String course) throws SQLException {
+    return getTrendingNotesBySchoolOrCourse(pool, Recency.TODAY, school, course);
+  }
+
+  /** Gets trending notes this week filtered by school or course */
+  public Object[][] getTrendingNotesThisWeekBySchoolOrCourse(DataSource pool, String school, String course) throws SQLException {
+    return getTrendingNotesBySchoolOrCourse(pool, Recency.THIS_WEEK, school, course);
+  }
+
+  /** Gets trending notes this month filtered by school or course */
+  public Object[][] getTrendingNotesThisMonthBySchoolOrCourse(DataSource pool, String school, String course) throws SQLException {
+    return getTrendingNotesBySchoolOrCourse(pool, Recency.THIS_MONTH, school, course);
+  }
+
+  /** Gets all-time trending notes filtered by school or course */
+  public Object[][] getTrendingNotesAllTimeBySchoolOrCourse(DataSource pool, String school, String course) throws SQLException {
+    return getTrendingNotesBySchoolOrCourse(pool, Recency.ALL_TIME, school, course);
+  }
+
+  /** Gets trending notes based on number of favorites in a given timespan, filtered by school and/or course */
+  private Object[][] getTrendingNotesBySchoolOrCourse(DataSource pool, Recency recency, String school, String course) throws SQLException {
+    String filterStmt = getStatementToFilterBySchoolOrCourse(school, course);
+    Date date = getDateBasedOnRecency(recency);
+    List<Object> notes = new ArrayList<>();
+    try (Connection conn = pool.getConnection()) {
+      String stmt = getTrendingNotesStatement(filterStmt);
+      try (PreparedStatement selectStmt = conn.prepareStatement(stmt)) {
+        selectStmt.setDate(1, date);
+        ResultSet rs = selectStmt.executeQuery();
+        while (rs.next()) {
+          Note thisNote = constructNoteFromSqlResult(pool, rs);
+          Long numFavoritesInTimespan = rs.getLong(NUM_FAVORITES_IN_TIMESPAN_ID);
+          notes.add(new Object[] {thisNote, numFavoritesInTimespan});
+        }
+        rs.close();
+        return notes.toArray(new Object[0][0]);
+      }
+    }
+  }
+
+  /** Gets a sql where clause that filters by school or course if school or course is not null */
+  private String getStatementToFilterBySchoolOrCourse(String school, String course) {
+    String filterStmt = "WHERE 1=1 ";
+    if (school != null && !school.isEmpty()) {
+      filterStmt += "AND `school`= '" + school + "' ";
+    }
+    if (course != null && !course.isEmpty()) {
+      filterStmt += "AND `course`= '" + course + "' ";
+    }
+    return filterStmt;
+  }
+
+  /** Gets trending notes based on number of favorites in a given timespan */
+  private Object[][] getTrendingNotes(DataSource pool, Recency recency) throws SQLException {
```

Comment:
I don't think this is necessary-- you can simply modify the helper method for getting the filter statement to return an empty string if there is no school or course passed in so you can avoid repeating all of this logic.

Similarly, I also feel like it would be possible to reduce the number of methods above by simply passing null for school/course so you don't have to kind of have duplicate wrappers to get trending notes per time period

Reply:
You're right!! I didn't catch this at all but the null checks for school and course are already being done in the filter statement and should actually be usable as is in a single getTrendingNotes method that can take a null school or course. I'll get rid of all the wrappers and make a single method that takes school and course with the usage that if school or course are null, then there's no real filter being applied.